[KOGITO-9982] Add persistence to SonataFlow CRD (#529)
apache/incubator-kie-kogito-docs: serverlessworkflow/modules/ROOT/pages/cloud/operator/using-persistence.adoc

= Building a Custom Development Image
:compat-mode!:
// Metadata:
:description: Using persistence in the workflow instance to store its context
:keywords: sonataflow, workflow, serverless, operator, kubernetes, persistence

This document describes how to configure a SonataFlow instance to use persistence to store the flow's context in a relational database.

== Configuring the SonataFlow CR to use persistence

> Kubernetes's pods are stateless by definition. In some scenarios this can be a challenge for workloads that require to maintain the status of 

Reviewer: ```suggestion
Kubernetes's pods are stateless by definition. In some scenarios, this can be a challenge for workloads that require maintaining the status of 
```

the application regardless of the pod's lifecycle. In the case of SonataFlow, the context of the workflow is lost when the pod restarts.

> If your workflow requires to recover from such scenarios, you need to make these additions to your Workflow CR:

Reviewer: ```suggestion
If your workflow requires recovery from such scenarios, you must make these additions to your Workflow CR:
```

Use the `persistence` field in the `SonataFlow` workflow spec to define the database service located in the same cluster.
There are 2 ways to accomplish this:

* Using the Platform CR's defined persistence

> When the Platform CR is deployed with its persistence spec populated it enables workflows to leverage on its configuration to populate the persistence

Reviewer: ```suggestion
When the Platform CR is deployed with its persistence spec populated it enables workflows to leverage its configuration to populate the persistence
```

properties in the workflows. 

[source,yaml,subs="attributes+"]
----
apiVersion: sonataflow.org/v1alpha08
kind: SonataFlowPlatform
metadata:
  name: sonataflow-platform
spec:
  persistence:
    postgresql:
      secretRef:
        name: postgres-secrets
        userKey: POSTGRES_USER
        passwordKey: POSTGRES_PASSWORD
      serviceRef:
        name: postgres
        port: 5432
        databaseName: sonataflow
        databaseSchema: shared
  build:
    config:
      strategyOptions:
        KanikoBuildCacheEnabled: "true"
---

> The values of `POSTGRES_USER` and `POSTGRES_PASSWORD` are the keys in the secret that contains the credentials to connect to the postgreSQL instance. 

Reviewer: ```suggestion
The values of `POSTGRES_USER` and `POSTGRES_PASSWORD` are the secret keys containing the credentials to connect to the PostgreSQL instance. 
```

Author: I'm not sure because this change conveys that the keys are a secret, when in fact the keys are not a secret but are stored in an object of kind `secret`. It's a kubernetes [terminology](https://kubernetes.io/docs/concepts/configuration/secret/).

Reviewer 2: imho we can add the link:
`...are the keys in the [secret](https://kubernetes.io/docs/concepts/configuration/secret/)`

The SonataFlow Workflow CR 
is defined with its `Persistence` field defined as empty.

[source,yaml,subs="attributes+"]
apiVersion: sonataflow.org/v1alpha08
kind: SonataFlow
metadata:
  name: callbackstatetimeouts
  annotations:
    sonataflow.org/description: Callback State Timeouts Example k8s
    sonataflow.org/version: 0.0.1
spec:
  persistence: {}
  ...
---

This configuration signals the operator that the workflow requires persistence and that it expects its configuration to be populated accordingly. 

> The operator will add the revelant JDBC properties in the `application.properties` 

Reviewer: ```suggestion
The operator will add the relevant JDBC properties in the `application.properties` 
```

generated and as part of the pod´s environment so that it can connect to the persistence service defined in the `Platform` CR.

> Note that currently only PostgreSQL is the only persistence supported.

Reviewer 3: ```suggestion
[NOTE]
====
Currently, PostgreSQL is the only persistence supported.
====
```

> * Using the custom defined persistence in the SonataFlow CR

Reviewer 3: ```suggestion
* Using the custom defined persistence in the `SonataFlow` CR
```

> Alternatively you can define a dedicated configuration in the SonataFlow CR instance using the same schema format found in the Platform CRD:

Reviewer: ```suggestion
Alternatively, you can define a dedicated configuration in the SonataFlow CR instance using the same schema format found in the Platform CRD:
```

> Alternatively, you can define a dedicated configuration in the SonataFlow CR instance using the same schema format found in the Platform CRD:

Reviewer 3: ```suggestion
Alternatively, you can define a dedicated configuration in the `SonataFlow` CR instance using the same schema format found in the Platform CRD:
```
